File: algorithms/BackTracking/permutations.py

def permute(nums):
    
    #print('nums ', len(nums))
    #print('arr ', len(arr))
    def per(nums, arr, ans):
        if not nums:
            if len(arr) == 3:
                print('arr ', arr)
                ans.extend(arr)
            return

        for i in nums:
            newNums = list(filter(lambda x: x != i, nums))
            arr.append(i)
            per(newNums, arr, ans)
            arr.pop()

    ans = []
    if not nums:
        return ans
    per(nums, [], ans)
    return ans

    


print(permute([4, 2, 3]))

File: algorithms/neetcode/000/BackTracking/permutations.js

var permute = function(nums, set=[], answers=[]) {
	if (!nums.length) answers.push([...set]);

	for(let i=0; i < nums.length; i++){
		const newNums = nums.filter((n, index) => index != i);
		set.push(nums[i]);
		permute(newNums, set, answers);
		set.pop();
	}
	return answers;
};

console.log(permute([1,2,3]))

const results = permute(['b1','b2','g'])

results.forEach(function(result){
	if(JSON.stringify(['b1','g','b2']) !== JSON.stringify(result) &&
		JSON.stringify(['b2','g','b1']) !== JSON.stringify(result))
	{
	console.log(result)
	}
});
